### PR TITLE
Show repository snapshot list

### LINKED ISSUE #319: Show repository snapshot list
- [x] Add api types
- [x] Create Repository crd
- [x] Implement registry for Snapshot
- [ ] Update docs/installer
 
```
repository.stash.appscode.com / v1alpha1
RepoSnapshpot

apis/restic.stash.appscode.com/v1alpha1/namespaces/<ns>/reposnaphsots/<RESTIC>/<>
?
workload_kind=
workload_name=
nodename=
pod_ordinal=

Workload        NodeName    Ordinal   SnapshotID  <snapshot timestamp>  path
deploy/<name>
ds/<name>       <node>


type Snapshot struct {
	TypeMeta
	ObjectMeta
	Data  []Repository
}

type Repository struct {
	Path
	workload_kind
	workload_name
	nodename
	pod_ordinal

	History []SnapshotData
}
```

### PR DESCRIPTION
**Pull request by emruz-hossain**
This PR fixes #319 
Code:
- [x] Implement List method for snapshots
- [x] Implement Get method for snapshots
- [x] Implement Delete method for snapshots

Test:
- [x] Add test for snapshots

Docs:
- [x] Add doc explaining uses of snapshot

### CODE REVIEW

File: pkg/cli/restic.go [hunk 1]
```diff
@@ -47,15 +47,32 @@ type Snapshot struct {
...
 }
 
-func (w *ResticWrapper) ListSnapshots() ([]Snapshot, error) {
+func (w *ResticWrapper) ListSnapshots(snapshotIDs []string) ([]Snapshot, error) {
 	result := make([]Snapshot, 0)
-	args := w.appendCacheDirFlag([]interface{}{"snapshots", "--json"})
+	args := w.appendCacheDirFlag([]interface{}{"snapshots", "--json", "--quiet"})
 	args = w.appendCaCertFlag(args)
 
+	if snapshotIDs != nil {
```
Comment (diptadas) [inline pkg/cli/restic.go]: don't need to perform a nil check here
https://caretdashcaret.com/2017/04/16/iterating-over-nil-slices-in-go/

Comment (emruz-hossain) [reply pkg/cli/restic.go]: Ok. refactored.

File: pkg/registry/snapshot/snapshot-util.go [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+	}
+
+	response, err := r.execCommandOnPod(pod, command)
+	if err != nil {
+		return nil, err
+	}
+
+	return response, nil
+}
+func extractDataFromRepositoryLabel(labels map[string]string) (string, string, string, string, error) {
```
Comment (diptadas) [inline pkg/registry/snapshot/snapshot-util.go]: better to use named return value here

Comment (diptadas) [inline pkg/registry/snapshot/snapshot-util.go]: better to use named return value here

Comment (tamalsaha) [reply pkg/registry/snapshot/snapshot-util.go]: Use struct

**Comment by tamalsaha** [comment]
- [ ] Remove ObjectRef

File: test/e2e/framework/minio_server.go [hunk 3]
```diff
@@ -46,6 +46,9 @@ func (fi *Invocation) CreateMinioServer() (string, error) {
 
 	//creating deployment for minio server
 	mdeploy = fi.DeploymentForMinioServer()
+	if !tls { // if tls not enabled then don't mount secret for cacerts
+		mdeploy.Spec.Template.Spec.Containers[0].VolumeMounts = mdeploy.Spec.Template.Spec.Containers[0].VolumeMounts[0:1]
```
Comment (tamalsaha) [inline test/e2e/framework/minio_server.go]: Remove volume by name

File: test/e2e/framework/util.go [hunk 4]
```diff
@@ -37,7 +37,7 @@ func (f *Framework) EventualEvent(meta metav1.ObjectMeta) GomegaAsyncAssertion {
 func (f *Framework) EventualWarning(meta metav1.ObjectMeta) GomegaAsyncAssertion {
 	return Eventually(func() []core.Event {
 		fieldSelector := fields.SelectorFromSet(fields.Set{
-			"involvedObject.kind":      "Repository",
+			"involvedObject.kind":      "Restic",
```
Comment (tamalsaha) [inline test/e2e/framework/util.go]: Is this correct?

File: test/e2e/snapshots_test.go [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+			shouldHaveSidecar()
+
+			By("Waiting for Repository CRD")
+			shouldHaveRepositoryCRD()
+
+			By("Waiting for backup to complete")
+			shouldBackupComplete()
+
+			By("Listing all snapshots")
+			snapshots, err := f.StashClient.RepositoriesV1alpha1().Snapshots(f.Namespace()).List(metav1.ListOptions{})
```
Comment (tamalsaha) [inline test/e2e/snapshots_test.go]: Need to filter for this particular test

File: pkg/registry/snapshot/snapshot.go [hunk 6]
```diff
@@ -38,20 +45,135 @@ func (r *REST) GroupVersionKind(containingGV schema.GroupVersion) schema.GroupVe
...
+	if !ok {
+		return nil, errors.New("missing namespace")
+	}
+	if len(name) < 9 {
+		return nil, errors.New("invalid snapshot name")
+	}
+	snapshotIDs := make([]string, 0)
+	snapshotIDs = append(snapshotIDs, name[len(name)-8:])
+
+	repoName := strings.TrimSuffix(name, name[len(name)-9:])
```
Comment (tamalsaha) [inline pkg/registry/snapshot/snapshot.go]: Need to check index length

File: pkg/registry/snapshot/snapshot.go [hunk 7]
```diff
@@ -38,20 +45,135 @@ func (r *REST) GroupVersionKind(containingGV schema.GroupVersion) schema.GroupVe
...
+	}
+	var labelSelector string
+	if options.LabelSelector != nil {
+		labelSelector = options.LabelSelector.String()
+	}
+	repos := make([]v1alpha1.Repository, 0)
+
+	// if the request is for snapshots of specific repositories then use Get method to get those repositories
+	// otherwise use labelSelector to list all repositories that match the selectors.
+	if strings.Contains(labelSelector, "repository=") {
```
Comment (tamalsaha) [inline pkg/registry/snapshot/snapshot.go]: Convert to `Requirements` and test using that https://github.com/appscode/stash/blob/master/vendor/k8s.io/apimachinery/pkg/labels/selector.go#L53

File: pkg/registry/snapshot/snapshot.go [hunk 8]
```diff
@@ -38,20 +44,124 @@ func (r *REST) GroupVersionKind(containingGV schema.GroupVersion) schema.GroupVe
...
+		}
+	}
+
+	repoSelector := labels.Everything().Add(requirements...)
+	var selectedRepos []v1alpha1.Repository
+	for _, r := range repositories.Items {
+		repoLabels := make(map[string]string)
+		repoLabels = r.Labels
+		repoLabels["repository"] = r.Name
+		if repoSelector.Matches(labels.Set(repoLabels)) {
```
Comment (tamalsaha) [inline pkg/registry/snapshot/snapshot.go]: Just use `options.LabelSelector ` directly.

**Pull request merged**